src/core/sys/windows/stat.d

```diff
@@ -30,20 +30,40 @@ int S_ISDIR(int m)  { return (m & S_IFMT) == S_IFDIR; }
 int S_ISCHR(int m)  { return (m & S_IFMT) == S_IFCHR; }
 }
 
-struct struct_stat
+version (CRuntime_DigitalMars)
 {
-    short st_dev;
-    ushort st_ino;
-    ushort st_mode;
-    short st_nlink;
-    ushort st_uid;
-    ushort st_gid;
-    short st_rdev;
-    short dummy;
-    int st_size;
-    int st_atime;
-    int st_mtime;
-    int st_ctime;
+    struct struct_stat
+    {
+        short st_dev;
+        ushort st_ino;
+        ushort st_mode;
+        short st_nlink;
+        ushort st_uid;
+        ushort st_gid;
+        short st_rdev;
+        short dummy;
+        int st_size;
+        int st_atime;
+        int st_mtime;
+        int st_ctime;
+    }
+}
+else version (CRuntime_Microsoft)
+{
+    struct struct_stat
+    {
+        uint st_dev;
+        ushort st_ino;
+        ushort st_mode;
+        short st_nlink;
+        short st_uid;
+        short st_gid;
+        uint st_rdev;
+        int st_size;
+        int st_atime;
+        int st_mtime;
+        int st_ctime;
```

Comment:
The 3 timestamps are 64-bit `long` (`time_t`) for VS 2019 x64 / Win10 SDK.

Reply:
i've added pragmas to make the functions use `_stat32`, `_fstat32`, `_wstat32` which should use 32-bit time_t (documented [here](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/stat-functions?view=msvc-170#time-type-and-file-length-type-variations-of-_stat)) regardless of the C runtime's default, is this ok?

druntime has a 32-bit time_t defined for windows in `core.sys.windows.stdc.time` so fixing that to be 64-bit may be out of scope for this PR

Reply:
> druntime has a 32-bit time_t defined for windows in core.sys.windows.stdc.time so fixing that to be 64-bit may be out of scope for this PR

Ouch. :/ - I agree, out of scope here, but you're not using `time_t`, `_dev_t`, etc. here.

> i've added pragmas […] is this ok?

Nope, not IMO. ;) - We're mainly trying to be compatible with C code (and being able to link to C libs), and my headers even have a
```
#if defined _USE_32BIT_TIME_T && defined _WIN64
    #error You cannot use 32-bit time_t (_USE_32BIT_TIME_T) with _WIN64
#endif
```

I guess the best way forward is to use `time_t`, even if it's currently wrong, and to file an issue about that.

---

This little module doesn't even declare the 64-bit versions (with 64-bit file length), which in times of terabyte disks is… suboptimal. ;)

Reply:
the structs now use `time_t` but i've kept the pragmas because otherwise it would overflow trying to write the 64-bit times on the C runtime you're using

there seems to be an open issue about 64-bit time_t already: https://issues.dlang.org/show_bug.cgi?id=21134

Reply:
Hmm, that makes it harder to fix it once `time_t` is fixed, especially as there's no comment at all explaining why the pragmas are used. So if you really want to keep them, I suggest adding a comment linking to the time_t issue, and adding a comment in that issue linking to this PR.

Reply:
added a comment for the pragmas

> The 3 timestamps are 64-bit long (time_t) for VS 2019 x64 / Win10 SDK.

just checking, did you test this using the new struct in D or was this about its definition in C? i assumed from your comment that plain `stat()` has been changed to use 64-bit time values in the C runtime you're using (but i could've been wrong)

the original PR without the pragmas worked for me with the copy of `msvcr120` that dmd comes with, plain `stat()` there returns 32-bit time values (and file size) when compiled as both 64-bit and 32-bit

you could test it with something like this:

```D
alias time_t = int;
struct struct_stat
{
    uint st_dev;
    ushort st_ino;
    ushort st_mode;
    short st_nlink;
    short st_uid;
    short st_gid;
    uint st_rdev;
    int st_size;
    time_t st_atime;
    time_t st_mtime;
    time_t st_ctime;
}
extern(C) int stat(const(char)*, struct_stat *);
void main()
{
    struct X { struct_stat sb; ubyte[16] junk; }
    X x;
    assert(stat(".", &x.sb) == 0, "stat error");
    assert(x.junk == (ubyte[16]).init); // should fail if the struct was wrong
}
```

if the default for the plainly named versions really has changed, then all the other C library functions in druntime that deal with time_t might need fixing to work with `VS 2019 x64 / Win10 SDK` as well

Reply:
Oh, interesting. First off, your example runs fine. And that's because my `sys/stat.h` uses:
```C
    #ifdef _USE_32BIT_TIME_T

        static __inline int __CRTDECL fstat(int const _FileHandle, struct stat* const _Stat)
        {
            _STATIC_ASSERT(sizeof(struct stat) == sizeof(struct _stat32));
            return _fstat32(_FileHandle, (struct _stat32*)_Stat);
        }

        static __inline int __CRTDECL stat(char const* const _FileName, struct stat* const _Stat)
        {
            _STATIC_ASSERT(sizeof(struct stat) == sizeof(struct _stat32));
            return _stat32(_FileName, (struct _stat32*)_Stat);
        }

    #else

        static __inline int __CRTDECL fstat(int const _FileHandle, struct stat* const _Stat)
        {
            _STATIC_ASSERT(sizeof(struct stat) == sizeof(struct _stat64i32));
            return _fstat64i32(_FileHandle, (struct _stat64i32*)_Stat);
        }
        static __inline int __CRTDECL stat(char const* const _FileName, struct stat* const _Stat)
        {
            _STATIC_ASSERT(sizeof(struct stat) == sizeof(struct _stat64i32));
            return _stat64i32(_FileName, (struct _stat64i32*)_Stat);
        }

    #endif
```

> with the copy of msvcr120 that dmd comes with

Note that those VS 2013 libs are only used as fallbacks in case the compiler cannot find a Visual C++ installation.
So apparently they preserve backwards compatibility by defining **inline** wrappers, and the actual `stat` C symbol is still the old one with 32-bit timestamps.

The proper solution is probably to make `stat` an alias to `_stat64i32` accordingly, and define the struct with a proper `time_t`. But as you've mentioned, there'll probably be similar trouble for other functions dealing with `time_t`.

Reply:
> Oh, interesting. First off, your example runs fine.

thanks for checking, i've updated the comment and removed the pragmas since they don't seem to be necessary with the current 32-bit time_t